Bugfix/4.4.3

tomereli opened the pull request:
agent: refactor M2 teardown handling

According to thethe MultiAP specification:
"If a Multi-AP Controller receives a WSC TLV containing an M1, then it
shall respond within one second with either 1) one or more M2s for each
BSS to be configured on the radio identified in the AP Radio Identifier
TLV or 2) one M2 with bit 4 (Tear down bit) set to 1 in the Multi-AP
Extension subelement to indicate that zero BSS are to be configured on
the radio identified in the AP Radio Identifier TLV per section 17.1.3."

In practice, it seems that some controllers ignore this and send
multiple M2 TLVs for the same RUID with teardown bit set.

In addition, some controllers simply sends an empty M2 (with invalid attributes)
instead of explicitly setting teardown.

Therefore, refactor the M2 teardown handling in the agent as follows:
- Invalid M2 TLV - teardown
- Teardown bit set but more than one M2 - teardown

Also update test_flows.sh accordingly since the debug messages have
changed.

Fixes #360 

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: wsc: update configData to support teardown

Preparative commit.
If teardown is set in configData, all attributes except for vendor
extension which holds the teardown bit are optional (and irrelevant).
Therefore, update the getters to not fail if these attributes are
missing.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: refactor M2 teardown handling

According to thethe MultiAP specification:
"If a Multi-AP Controller receives a WSC TLV containing an M1, then it
shall respond within one second with either 1) one or more M2s for each
BSS to be configured on the radio identified in the AP Radio Identifier
TLV or 2) one M2 with bit 4 (Tear down bit) set to 1 in the Multi-AP
Extension subelement to indicate that zero BSS are to be configured on
the radio identified in the AP Radio Identifier TLV per section 17.1.3."

In practice, it seems that some controllers ignore this and send
multiple M2 TLVs for the same RUID with teardown bit set.

In addition, some controllers simply sends an empty M2 (with invalid attributes)
instead of explicitly setting teardown.

Therefore, refactor the M2 teardown handling in the agent as follows:
- Invalid M2 TLV - teardown
- Teardown bit set but more than one M2 - teardown

Also update test_flows.sh accordingly since the debug messages have
changed.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3965,9 +3965,16 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
+    config.ssid        = config_data->ssid();
+    config.bss_type    = static_cast<WSC::eWscVendorExtSubelementBssType>(config_data->bss_type());
     // Swap to network byte order for KWA HMAC calculation
+    // from this point config data is not readable!
```
Good idea to add this!

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4104,13 +4091,7 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
-        return false;
-    }
-
+    std::list<WSC::configData::config> configs;
```
std::list should only be used for lists where you insert in the middle relatively frequently. Use std::vector [SL.con.2](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rsl-vector)

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3965,9 +3965,16 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
         LOG(ERROR) << "invalid config data length";
         return false;
     }
+    // Update VAP configuration
+    config.auth_type   = config_data->auth_type();
+    config.encr_type   = config_data->encr_type();
+    config.bssid       = config_data->bssid();
+    config.network_key = config_data->network_key();
+    config.ssid        = config_data->ssid();
+    config.bss_type    = static_cast<WSC::eWscVendorExtSubelementBssType>(config_data->bss_type());
```
I would have liked this refactoring to be in a separate commit. (nitpick)

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4121,32 +4102,49 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
-                   << ", encryption_type: " << int(credentials->encryption_type_attr().data)
-                   << (fronthaul ? " fronthaul" : "") << (backhaul ? " backhaul" : "")
-                   << (teardown ? " teardown" : "");
+        WSC::configData::config config_;
```
Why `config_` and not just `config`?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4121,32 +4102,49 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
-                   << (teardown ? " teardown" : "");
+        WSC::configData::config config_;
+        if (!autoconfig_wsc_parse_m2_encrypted_settings(m2, authkey, keywrapkey, config_)) {
+            LOG(ERROR) << "Invalid config data, tear down radio";
```
I think we'll need to revisit this at some point. From a security perspective, if authentication fails, we should do nothing instead of tearing down. If config data is invalid it's a bit uncertain what we should do.

tomereli replied:
You are absolutely right - I rather do this right now than postpone it.
I changed the way we parse config data so that there is a difference between encryption/authentication failures - in which we ignore the config data, and empty config_data which is also invalid according to Table 20 in the WSC specification but Marvell controller sends it none the less.

arnout commented on tests/test_flows.sh:
```diff
@@ -123,8 +123,6 @@ test_initial_ap_config() {
     check_log ${REPEATER1} agent_wlan2 "WSC Global authentication success"
     check_log ${REPEATER1} agent_wlan0 "KWA (Key Wrap Auth) success"
     check_log ${REPEATER1} agent_wlan2 "KWA (Key Wrap Auth) success"
-    check_log ${REPEATER1} agent_wlan0 "Controller configuration (WSC M2 Encrypted Settings)"
-    check_log ${REPEATER1} agent_wlan2 "Controller configuration (WSC M2 Encrypted Settings)"
```
Just removing them is not nice. Can't we check something here?

tomereli replied:
Fixed and brought it back.

rmelotte reviewed (approved):
:+1:

Commit:
tlvf: configData::config remove "NA"

Preparative commit.
Remove "NA" from the struct initialization. Instead, use empty string.
This will be used to mark empty or invalid configs on parse failures.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
